################################################################################
SHORT DESCRIPTION: 
################################################################################
XSA-183 (http://xenbits.xen.org/xsa/advisory-183.html)
Missing SMAP whitelisting in 32-bit exception / event delivery.

################################################################################
LONG DESCRIPTION: 
################################################################################
Source: http://xenbits.xen.org/xsa/advisory-183.html
Patches: xsa183-4.6.patch

Supervisor Mode Access Prevention is a hardware feature designed to make an
Operating System more robust, by raising a pagefault rather than accidentally
following a pointer into userspace.  However, legitimate accesses into
userspace require whitelisting, and the exception delivery mechanism for 32bit
PV guests wasn't whitelisted.

CVE-2016-6259

################################################################################
PATCHES 
################################################################################
Index: xen-4.6.1/xen/arch/x86/x86_64/compat/entry.S
===================================================================
--- xen-4.6.1.orig/xen/arch/x86/x86_64/compat/entry.S	2016-08-31 16:33:27.238877183 +0200
+++ xen-4.6.1/xen/arch/x86/x86_64/compat/entry.S	2016-08-31 16:38:57.585698435 +0200
@@ -280,6 +280,7 @@
 compat_create_bounce_frame:
         ASSERT_INTERRUPTS_ENABLED
         mov   %fs,%edi
+        ASM_STAC
         testb $2,UREGS_cs+8(%rsp)
         jz    1f
         /* Push new frame at registered guest-OS stack base. */
@@ -333,6 +334,7 @@
         movl  %ds,%eax
 .Lft12: movl  %eax,%fs:0*4(%rsi)        # DS
 UNLIKELY_END(compat_bounce_failsafe)
+        ASM_CLAC
         /* Rewrite our stack frame and return to guest-OS mode. */
         /* IA32 Ref. Vol. 3: TF, VM, RF and NT flags are cleared on trap. */
         andl  $~(X86_EFLAGS_VM|X86_EFLAGS_RF|\
@@ -378,6 +380,7 @@
         addl  $4,%esi
 compat_crash_page_fault:
 .Lft14: mov   %edi,%fs
+        ASM_CLAC
         movl  %esi,%edi
         call  show_page_walk
         jmp   dom_crash_sync_extable
Index: xen-4.6.1/xen/arch/x86/x86_64/entry.S
===================================================================
--- xen-4.6.1.orig/xen/arch/x86/x86_64/entry.S	2016-08-31 16:33:27.262210292 +0200
+++ xen-4.6.1/xen/arch/x86/x86_64/entry.S	2016-08-31 16:38:57.682364172 +0200
@@ -462,9 +462,11 @@
 domain_crash_page_fault_8:
         addq  $8,%rsi
 domain_crash_page_fault:
+        ASM_CLAC
         movq  %rsi,%rdi
         call  show_page_walk
 ENTRY(dom_crash_sync_extable)
+        ASM_CLAC
         # Get out of the guest-save area of the stack.
         GET_STACK_BASE(%rax)
         leaq  STACK_CPUINFO_FIELD(guest_cpu_user_regs)(%rax),%rsp
